### ISSUE TITLE
Running out of memory at the catlas building phase

### ISSUE THREAD

**Issue by Laura-Alex** [spacegraphcats/spacegraphcats#243]
Hello,

I gave spacegraphcats another go, and tried to run bits of the code separately so I can understand where things go wrong. This time it's running out of memory at the catlas building phase (similar to #57, maybe?).

Running bcalm looks fine, although it did require a lot of cores (16+).
I tried to run 'spacegraphcats config.yaml build' next, and it seems to build the right folder (SRR2_k21_r1, with my catlas name being SRR2) and the right files (cdbg.gxt  contigs.fa.gz  contigs.fa.gz.info.csv  contigs.fa.gz.sig). However, it fails at the catlas building step. I've tried re-running 'build', hoping it would just use the already made files, but no dice.

The error message:

```
--------
details!
 snakefile: /[...]/home/anaconda3/envs/space/lib/python3.6/site-packages/spacegraphcats/./conf/Snakefile
 config: config.yaml
 targets: ['build']
--------
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1
Rules claiming more threads will be scaled down.
Job counts:
 count jobs
 1 build
 1 build_catlas
 1 make_contigs_kmer_index
 3

[Fri Oct  4 09:34:55 2019]
rule build_catlas:
    input: SRR2_k21_r1/cdbg.gxt, SRR2_k21_r1/contigs.fa.gz
    output: SRR2_k21_r1/first_doms.txt, SRR2_k21_r1/catlas.csv, SRR2_k21_r1/commands.log
    jobid: 1

/[...]/home/anaconda3/envs/space/bin/python -m spacegraphcats.catlas.catlas --no_checkpoint SRR2_k21_r1 1
=>> PBS: job killed: mem 101148464kb exceeded limit 100663296kb
```

That's 96GB it's not happy with...

Things that I changed from last time in order to try to make things work: I used fasta input files instead of fastq, and I used this guideline for the install (https://github.com/spacegraphcats/spacegraphcats-dory-example). However I still used khmer's trim_low_abund.py. The original fastq file was 4.9GB, the trimmed file was 4.4, the fasta was 2.5, the bcalm result 'bcalm.SRR2.k21.unitigs.fa' is 6.4GB, and some other files are as follows:
2.5G cdbg.gxt
726M contigs.fa.gz
1.7G contigs.fa.gz.info.csv
14M contigs.fa.gz.sig

Please help?

**Comment by ctb** [comment]
Hi Laura-Alex, thanks for following through!

yes, out of memory errors. Two ideas --

1) are you error trimming with khmer? I, umm, just realized that we did not
   put that in *any* of the tutorials, so my apologies... The command is

   trim-low-abund.py -C 3 -Z 18 -M 20e9 -V {input} -o {output} --gzip

where you replace {input} with all your input FASTQ filenames, and {output} with your desired output filename (it will be a gzipped file). Note khmer and
the trim-low-abund script will already be installed.

This will remove low-abundance k-mers and could significantly decrease
the size of the cDBG.

2) make sure you are NOT specifying 'keep_graph_pendants: true' in the spacegraphcats config file.  By default it is keep_graph_pendants is false, so if you don't find it in your config file you are good!

best,
--titus

On Fri, Oct 04, 2019 at 03:29:45AM -0700, Laura-Alex wrote:
> Hello,
> 
> I gave spacegraphcats another go, and tried to run bits of the code separately so I can understand where things go wrong. This time it's running out of memory at the catlas building phase (similar to #57, maybe?).
> 
> Running bcalm looks fine, although it did require a lot of cores (16+).
> I tried to run 'spacegraphcats config.yaml build' next, and it seems to build the right folder (SRR2_k21_r1, with my catlas name being SRR2) and the right files (cdbg.gxt  contigs.fa.gz  contigs.fa.gz.info.csv  contigs.fa.gz.sig). However, it fails at the catlas building step. I've tried re-running 'build', hoping it would just use the already made files, but no dice.
> 
> The error message:
> 
> ```
> --------
> details!
>  snakefile: /[...]/home/anaconda3/envs/space/lib/python3.6/site-packages/spacegraphcats/./conf/Snakefile
>  config: config.yaml
>  targets: ['build']
> --------
> Building DAG of jobs...
> Using shell: /bin/bash
> Provided cores: 1
> Rules claiming more threads will be scaled down.
> Job counts:
>  count jobs
>  1 build
>  1 build_catlas
>  1 make_contigs_kmer_index
>  3
> 
> [Fri Oct  4 09:34:55 2019]
> rule build_catlas:
>     input: SRR2_k21_r1/cdbg.gxt, SRR2_k21_r1/contigs.fa.gz
>     output: SRR2_k21_r1/first_doms.txt, SRR2_k21_r1/catlas.csv, SRR2_k21_r1/commands.log
>     jobid: 1
> 
> /[...]/home/anaconda3/envs/space/bin/python -m spacegraphcats.catlas.catlas --no_checkpoint SRR2_k21_r1 1
> =>> PBS: job killed: mem 101148464kb exceeded limit 100663296kb
> ```
> 
> That's 96GB it's not happy with...
> 
> Things that I changed from last time in order to try to make things work: I used fasta input files instead of fastq, and I used this guideline for the install (https://github.com/spacegraphcats/spacegraphcats-dory-example). However I still used khmer's trim_low_abund.py. The original fastq file was 4.9GB, the trimmed file was 4.4, the fasta was 2.5, the bcalm result 'bcalm.SRR2.k21.unitigs.fa' is 6.4GB, and some other files are as follows:
> 2.5G cdbg.gxt
> 726M contigs.fa.gz
> 1.7G contigs.fa.gz.info.csv
> 14M contigs.fa.gz.sig
> 
> Please help?
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/spacegraphcats/spacegraphcats/issues/243
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Comment by Laura-Alex** [comment]
Thank you for your help.

Unfortunately, it didn't quite work...

I previously trimmed the data with khmer. The only thing I didn't do before is gzip it. I also removed 'keep_graph_pendants: true' from the config file, My exact code now is is:

`trim-low-abund.py -C 3 -Z 18 -M 48G -V -k 21 $PBS_O_WORKDIR/genome_in/SRR6462984.fastq -o $PBS_O_WORKDIR/genome_out/SRR6462984_sorted_2.fastq.gz --gzip`

Small note - the first sequences of my file, even after trimming, appear to be erroneous (eg. sequences composed just of Gs or Ns). Is this likely to affect downstream code?

The current error message:

`details!
        snakefile: /rds/general/user/laa412/home/anaconda3/envs/space/lib/python3.6/site-packages/spacegraphcats/./conf/Snakefile
        config: config.yaml
        targets: ['build']
--------
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1
Rules claiming more threads will be scaled down.
Job counts:
        count   jobs
        1       build
        1       build_catlas
        1       make_contigs_kmer_index
        3

[Wed Oct 16 12:31:32 2019]
rule build_catlas:
    input: SRR2_k21_r1/cdbg.gxt, SRR2_k21_r1/contigs.fa.gz
    output: SRR2_k21_r1/first_doms.txt, SRR2_k21_r1/catlas.csv, SRR2_k21_r1/commands.log
    jobid: 1
    resources: mem_mb=80000

/rds/general/user/laa412/home/anaconda3/envs/space/bin/python -m spacegraphcats.catlas.catlas --no_checkpoint SRR2_k21_r1 1
=>> PBS: job killed: mem 104664604kb exceeded limit 100663296kb`

I tried to give the code a limit of 80GB within the Snakefile, and requested 96GB from the cluster. It still went over.

I know that the downstream files can be created, because I submitted the job with a request for 480GB memory a while ago. At the very least, files first_doms.txt and catlas.csv existed then. (I intrerrupted that job because it felt wasteful/too impractical for the rest of my data given the waiting time).

Would you happen to have any insights?

**Comment by ctb** [comment]
Hi Laura-Alex, ok -- I think the only thing left to do is increase the
memory. I'm not sure why it requires so much - I'd guess your sample
is very diverse? - but there's nothing more we can do to reduce
memory requirements (or at least nothing we can do quickly).

best,
--titus

On Wed, Oct 16, 2019 at 06:46:54AM -0700, Laura-Alex wrote:
> Thank you for your help.
> 
> Unfortunately, it didn't quite work...
> 
> I previously trimmed the data with khmer. The only thing I didn't do before is gzip it. I also removed 'keep_graph_pendants: true' from the config file, My exact code now is is:
> 
> `trim-low-abund.py -C 3 -Z 18 -M 48G -V -k 21 $PBS_O_WORKDIR/genome_in/SRR6462984.fastq -o $PBS_O_WORKDIR/genome_out/SRR6462984_sorted_2.fastq.gz --gzip`
> 
> Small note - the first sequences of my file, even after trimming, appear to be erroneous (eg. sequences composed just of Gs or Ns). Is this likely to affect downstream code?
> 
> The current error message:
> 
> `details!
>         snakefile: /rds/general/user/laa412/home/anaconda3/envs/space/lib/python3.6/site-packages/spacegraphcats/./conf/Snakefile
>         config: config.yaml
>         targets: ['build']
> --------
> Building DAG of jobs...
> Using shell: /bin/bash
> Provided cores: 1
> Rules claiming more threads will be scaled down.
> Job counts:
>         count   jobs
>         1       build
>         1       build_catlas
>         1       make_contigs_kmer_index
>         3
> 
> [Wed Oct 16 12:31:32 2019]
> rule build_catlas:
>     input: SRR2_k21_r1/cdbg.gxt, SRR2_k21_r1/contigs.fa.gz
>     output: SRR2_k21_r1/first_doms.txt, SRR2_k21_r1/catlas.csv, SRR2_k21_r1/commands.log
>     jobid: 1
>     resources: mem_mb=80000
> 
> /rds/general/user/laa412/home/anaconda3/envs/space/bin/python -m spacegraphcats.catlas.catlas --no_checkpoint SRR2_k21_r1 1
> =>> PBS: job killed: mem 104664604kb exceeded limit 100663296kb`
> 
> I tried to give the code a limit of 80GB within the Snakefile, and requested 96GB from the cluster. It still went over.
> 
> I know that the downstream files can be created, because I submitted the job with a request for 480GB memory a while ago. At the very least, files first_doms.txt and catlas.csv existed then. (I intrerrupted that job because it felt wasteful/too impractical for the rest of my data given the waiting time).
> 
> Would you happen to have any insights?
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/spacegraphcats/spacegraphcats/issues/243#issuecomment-542709055
-- 
C. Titus Brown, ctbrown@ucdavis.edu

**Comment by Laura-Alex** [comment]
I had a look with Krona. It is indeed a far more diverse sample than the Hu et al. sample used in the spacegraphcats paper (https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR6462984&krona=on versus https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1976948&krona=on).

I was lucky enough to have my job immediatly submitted to the cluster. I thought it might be helpful to update on how things went, in case it makes things easier for me / anyone else who has this issue in the future.

spacegraphcats config.yaml build
...required 179GB (max) and an average of 2.79 cores, and approx 15 hours, 12 of which were dedicated to spacegraphcats.catlas.catlas. I actually got an error at that point, but apparently it wasn't enough to intrerrupt the program. The error:

/rds/general/user/[my username]/home/anaconda3/envs/space/bin/python -m spacegraphcats.catlas.catlas --no_checkpoint SRR2_k21_r1 1
fatal: Not a git repository (or any parent up to mount point /rds/general)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).

The following steps ('search' and 'extract_contigs'/'extract reads') were more reasonable resource- and time-wise. However, extract_contigs doesn't seem to do what it is supposed to do...

Extract_reads gave me 5 sequences. If aligned, and overlapped, I'd get a sequence 194 nucleotides long (which isn't any better than I could've done with Blast, but...I guess my organism might not be that abundant). Or perhaps changing the radius would help? I'm not sure I understand how the radius is supposed to work.

However, extract_contigs gave me multiple sequences...which were shorter than the reads. Something seems to have gone very wrong there.

Next I'll test a less diverse sample, to see if it works better.

**Comment by ctb** [comment]
thanks again for following through!

since the contigs (really unitigs) are from a compact De Bruijn graph,
with no additional processing, they may be as small as a single k-mer (k=31,
by default). So it's not unexpected that they can be smaller than the reads!

the radius relates to the fragmentation of the cDBG due to repeats,
sequencing errors, and polymorphism/strain variation. More graph fragmentation
from high polymorphism (and/or sequencing errors in low coverage organisms)
will result in the need to use higher radii to recover the neighborhood.

long story short, you might try r=5 or r=10 as well. they should work
in the same or less memory, and could yield more results.

**Comment by ctb** [comment]
One thing to add is that your graph may just be very fragmented due to low coverage or sequencing bias, in which case the only real solution to recovering more is to sequence more deeply.

**Comment by Laura-Alex** [comment]
Current update: radius of 10 ran out of memory using at first 240, then 480GB of memory. I haven't tried r5.

The alternative, less diverse dataset that I was considering testing turned out to be quite...large. (20GB after khmer) So, it's a bit impractical.

My current idea is to hopefully reduce some of that diversity/size by pre-filtering the sequences (eg. assigning the reads with Kraken, then using the reads with the right label for spacegraphcats). I wonder if anyone tried that before?

Either way, thank you for your help! Being able to understand where the issue is is really helpful.
